### PR TITLE
[mediaqueries-4] make explicit that `pointer`, `hover`, `any-pointer`, `any-hover` relate only to pointing devices

### PR DESCRIPTION
**Pull request by patrickhlauke**
Closes https://github.com/w3c/csswg-drafts/issues/967

### CODE REVIEW

**Comment by patrickhlauke** [comment]
While in #967 i proposed "pointer-based input mechanisms", i saw that the spec already used "pointing devices" as a phrase, so for consistency went with this one...

**Comment by inoas** [comment]
OT: So is iOS-Safari not behaving strict to the spec when it render elements you tap onto as having `:hover` triggered? Ref: http://codepen.io/anon/pen/jywdxM

**Comment by patrickhlauke** [comment]
yours is a question more generally about the spec, and not about something in this PR, but:

iOS uses heuristics to make `:hover` apply under certain conditions after a user has tapped the screen, but it's undisputable that the touchscreen itself does not support actual hovering, and not every tap (e.g. on non-actionable elements) results in a hover behavior either.

see https://drafts.csswg.org/mediaqueries-4/#hover in reference to `hover: none`:

> Pointing systems that can hover, but for which doing so is inconvenient and not part of the normal way they are used, also match this value

iOS' faked hovering would fall under this category.

In any case, this spec does not say either way when user agents should or shouldn't apply `:hover`...it specifies the `hover` media feature itself, which reflects the capabilities of the pointing device.

**Comment by inoas** [comment]
OT/Noise: Maybe it could be helpful for the future to specify that touch-only (without 3d-touch / pressing sensitivity at least) devices are no pointing devices (as there is no real way to distinquish between pointing and clicking/tapping)...

The spec you are correcting also speaks about multiple devices. It should trigger `:hover` for the pointing device of highest priority, not just any primary device (imagine touch+pointing on the same screen, touch being the primary device?). It could also explicitly exclude non-pointing devices such as (2d) touch.

However it would be great to have `:tap` or `:touch` to be able differentiate.

**Comment by patrickhlauke** [comment]
i think you're misunderstanding how the term "pointing device" is used here in the spec (and elsewhere, such as in https://w3c.github.io/pointerevents/). nothing to do with "pointing" as separate from potential activation.

If you do want to propose a new distinction/value for the media features here, i'd suggest opening a new issue.

Lastly, this spec does not deal with defining pseudo-classes like `:hover` or the hypothetical `:tap` / `:touch` you propose, so suggest actually going to https://drafts.csswg.org/selectors-4/ and proposing things there...

**Comment by inoas** [comment]
Thanks for clarification! and the link to pointerevents

**Comment by patrickhlauke** [comment]
as much as i hate pestering... @frivoal does this look ok to merge (based on your positive comment https://github.com/w3c/csswg-drafts/issues/967#issuecomment-274419675) ?

**Comment by patrickhlauke** [comment]
bumping this for @tabatkins' consideration as well...

**Comment by tabatkins** [comment]
Looks pretty reasonable to me. @frivoal, any objections to merging?

**Comment by frivoal** [comment]
This is probably good, but I wonder: this is fine for `any-*` but what if the input mechanism deemed primary is not a pointing device? Say, a device with a keyboard deemed primary, and a gamepad. Should `pointer` match `none` because that's the case of the keyboard, or should it match `coarse`, because within the pointing devices, the gamepad, by virtue of being the only one, is necessarily primary, even if overall it wouldn't be?

This is possibly enough of a corner case not to matter, but I am not sure.

File: mediaqueries/Overview.bs [hunk 1]
```diff
@@ -1597,18 +1597,24 @@ Interaction Media Features</h2>
...
-	Note: While this specification does not define how User Agents should decide what the “primary” input is, 
+	Note: While this specification does not define how User Agents should decide what the “primary” pointing device is, 
 	the expectation is that User Agents should make this determination 
 	by combining knowledge about the device/environment they are running on, 
-	the number and type of input mechanisms available, 
-	and a notion of which of these inputs is generally and/or currently being used. 
-	User Agents may also decide to dynamically change what type of input is deemed to be primary, 
+	the number and type of pointing devices available, 
+	and a notion of which of these is generally and/or currently being used. 
+	User Agents may also decide to dynamically change what type of pointing device is deemed to be primary,
```
Comment (patrickhlauke) [inline mediaqueries/Overview.bs]: could add an extra example right here: "In situations where the primary input mechanism for a device is not a pointing device, but there is a secondary – and less frequently used –  input that is a pointing devices, the user agent may decide to keep treating the non-pointing device as the primary (resulting in `pointer: none`), or they may opt to "promote" the secondary input to be identified as the primary pointing device (resulting in `pointer: fine` or `pointer: coarse`).

**Comment by patrickhlauke** [comment]
> what if the input mechanism deemed primary is not a pointing device? Say, a device with a keyboard deemed primary, and a gamepad. Should pointer match none because that's the case of the keyboard, or should it match coarse, because within the pointing devices, the gamepad, by virtue of being the only one, is necessarily primary, even if overall it wouldn't be?

i'd say that's the remit of the UA to decide in that case. i'd tend to lean towards "if the UA deems the keyboard to be the primary input mechanism, and the gamepad is only used very infrequently, then there's no primary pointer, but only an `any-` pointer input. what about adding an extra example to that first note? (just added it as a comment)

basically...up to the UA, and this spec won't say what's right and it's up to the UA. or is that too handwavy?

File: mediaqueries/Overview.bs (hunk 1 above)
Comment (frivoal) [inline mediaqueries/Overview.bs]: Sounds good. I'd shorten it to this:

In situations where the primary input mechanism for a device is not a pointing device, but there is a secondary – and less frequently used – input that is a pointing devices, the user agent may decide to treat the non-pointing device as the primary (resulting in `pointer: none`).

If you can update the pull request with that, I'll merge.

**Comment by patrickhlauke** [comment]
urgh, i seem to have completely messed up this pr by doing a rebase. let me see if i can fix that...

File: mediaqueries/Overview.bs (hunk 1 above)
Comment (patrickhlauke) [inline mediaqueries/Overview.bs]: Done. Agree the terse versions works just as well.

**Pull request merged**